go/private/tools.bzl

```diff
@@ -0,0 +1,69 @@
+# Copyright 2014 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+load("@io_bazel_rules_go//go/private:providers.bzl", "GoPath")
+load("@io_bazel_rules_go//go/private:common.bzl", "get_go_toolchain")
+
+def _go_vet_generate_impl(ctx):
+  print("""
+EXPERIMENTAL: the go_vet_test rule is still very experimental
+Please do not rely on it for production use, but feel free to use it and file issues
+""")
+  go_toolchain = get_go_toolchain(ctx)
+  script_file = ctx.new_file(ctx.label.name+".bash")
+  gopath = []
+  files = ctx.files.data
```

Comment:
In #726, they copied lists from `ctx.attr` (as `ctx.attr.linkopts[:]`). That seems to imply `ctx.files.data` will be immutable in the future.